submission/races.go

```diff
@@ -0,0 +1,227 @@
+	"github.com/google/certificate-transparency-go/ctpolicy"
+)
+
+const (
+	// PostBatchInterval is duration between parallel batch call and subsequent requests to Logs within group.
+	PostBatchInterval = time.Second
+)
+
+// submissionState holds outcome of a single-log submission.
+type submissionResult struct {
+	sct *ct.SignedCertificateTimestamp
+	err error
+}
+
+// safeSubmissionSet guards submission states for set of Logs.
+type safeSubmissionSet struct {
+	results map[string]*submissionResult
+	mu      sync.Mutex
+}
+
+// request includes empty submissionResult in the set, returns whether the entry wasn't requested (added) before.
+func (sub *safeSubmissionSet) request(logURL string) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	if sub.results[logURL] != nil {
+		return false
+	}
+	sub.results[logURL] = &submissionResult{}
+	return true
+}
+
+func (sub *safeSubmissionSet) has(logURL string) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	return sub.results[logURL] != nil
+}
+
+func (sub *safeSubmissionSet) sumSCTsFor(URLs map[string]bool) int {
+	var matches int
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	for url := range URLs {
+		if sub.results[url] != nil && sub.results[url].sct != nil {
+			matches++
+		}
+	}
+	return matches
+}
+
+func (sub *safeSubmissionSet) setResult(logURL string, sct *ct.SignedCertificateTimestamp, err error) {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	sub.results[logURL] = &submissionResult{sct: sct, err: err}
+}
+
+func (sub *safeSubmissionSet) collectSCTs() []*AssignedSCT {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	scts := []*AssignedSCT{}
+	for logURL, r := range sub.results {
+		if r != nil && r.sct != nil {
+			scts = append(scts, &AssignedSCT{LogURL: logURL, SCT: r.sct})
+		}
+	}
+	return scts
+}
+
+// postInterval calculates duration for consequent call. For first parallelStart calls duration is 0, while every next one gets additional dur interval.
+func postInterval(idx int, parallelStart int, dur time.Duration) time.Duration {
+	if idx < parallelStart {
+		return time.Duration(0)
+	}
+	return time.Duration(idx+1-parallelStart) * dur
+}
+
+// Submitter is interface wrapping Log-request-response cycle and any processing.
+type Submitter interface {
+	SubmitToLog(ctx context.Context, logURL string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error)
+}
+
+type groupState struct {
+	Name    string
+	Success bool
+}
+
+// groupRace shuffles logs within the group, submits avoiding duplicate-requests and collects responses. Upon (not)reaching required number of SCTs signals via completion channel.
+func groupRace(ctx context.Context, chain []ct.ASN1Cert, group *ctpolicy.LogGroupInfo, parallelStart int,
+	state *safeSubmissionSet, completion chan<- groupState, submitter Submitter) {
+	groupURLs := make([]string, 0, len(group.LogURLs))
+	for logURL := range group.LogURLs {
+		groupURLs = append(groupURLs, logURL)
+	}
+
+	type count struct{}
+	counter := make(chan count, len(groupURLs))
```

Comment:
`counter` looks it could be replace with a `sync.WaitGroup`:
```golang
	var wg sync.WaitGroup
	for i, urlNum := range rand.Perm(len(groupURLs)) {
		wg.Add(a)
		go func(i int, logURL string) {
			defer wg.Done()
			...
		}(i, groupURLs[urlNum])
	}
	// Wait until all per-group goroutines are done.
	wg.Wait()
	...
```
However, this will need the `Sleep` to be interruptible by a cancelled (sub-)context -- see below.

Reply:
Still need ability to check completion on each counter.

Reply:
This still looks like it should be a `sync.WaitGroup` to me.  Also, if you have an invariant that the inner goroutine must always do something on exit (whether it's `counter <- count{}` or `wg.Done()`), then use a `defer`.

Reply:
The logic has been updated since.